nep-26.mediawiki

```diff
@@ -0,0 +1,100 @@
+<pre>
+  NEP: 26
+  Title: Smart contract transfer callback for non-fungible tokens
+  Author: Erik Zhang <erik@neo.org>, Roman Khimov <roman@nspcc.ru>, Jaime Kindelan <jaime@red4sec.com>, Jimmy Liao <jimmy@r3e.network>
+  Type: Informational
+  Status: Accepted
+  Created: 2024-03-03
+  Requires: 11
+</pre>
+
+==Abstract==
+
+This proposal describes the callback method used by Neo smart contracts to
+process transfers of non-fungible [[nep-11.mediawiki|NEP-11]] tokens to their addresses.
+
+==Motivation==
+
+While the original [[nep-11.mediawiki|NEP-11]] already defines this method, it
+lacks some details relevant specifically to the receiver contract and it doesn't
+allow for easy reference. This standard can be used in <code>supportedstandards</code>
+section of [[nep-15.mediawiki|NEP-15 manifest]] or in any other situation to
+refer to contracts able to receive [[nep-11.mediawiki|NEP-11]] tokens (but not
+necessarily being [[nep-11.mediawiki|NEP-11]] themselves). Typical use cases
+for this functionality are: depositing/staking (with contract locking and
+managing tokens according to some rules) or exchange/minting (contract accepting
+transfer to itself, but returning some different token to the sender).
+
+==Specification==
+
+Smart contracts that need to receive and own [[nep-11.mediawiki|NEP-11]] tokens
+MUST implement <code>onNEP11Payment</code> method with the following signature:
+
+<pre>
+{
+  "name": "onNEP11Payment",
+  "parameters": [
+    {
+      "name": "from",
+      "type": "Hash160"
+    },
+    {
+      "name": "amount",
+      "type": "Integer"
+    },
+    {
+      "name": "tokenId",
+      "type": "ByteString"
+    },
+    {
+      "name": "data",
+      "type": "Any"
+    }
+  ],
+  "returntype": "Void"
+}
+</pre>
+
+A contract that doesn't implement this method will not be able to receive
```

Comment:
I found this paragraph a little confusing to me. Maybe re-write some parts of it. I will suggest. Fell free to accept, change or modify.

Reply:
> A contract that doesn't implement this method will not be able to receive [[nep-11.mediawiki|NEP-11]] tokens (transfers to its address will fail). 
Will fail or ABORT OR ABOTMSG OR ASSERT OR ASSETMSG?

>  This method is called by [[nep-11.mediawiki|NEP-11]] contracts which tokens are being
transferred. 
As written above, other NON-NEP-11 Contracts can have this

>A particular token can be obtained via the <code>System.Runtime.GetCallingScriptHash</code>
system call. 
Improve this sentence and describe with more detail. It is a little vague about a particular token can be obtained.
Explain about it a little more, let's say A particular token hash can be obtained from the transaction that is calling the contract. 

> Contracts SHOULD check for this hash to match their expectations
like allowed/accepted tokens, processing this call means that contract trusts
the caller to implement [[nep-11.mediawiki|NEP-11]] correctly.

Transfer with multiple tokens?

Reply:
> Will fail or ABORT OR ABOTMSG OR ASSERT OR ASSETMSG?

It's contract-specific, we just know it will fail the execution per NEP-11/NEP-17.

> As written above, other NON-NEP-11 Contracts can have this

You mean non-NEP-11 contracts doing `onNEPXXPayment` invocations? That's what `processing this call means that contract trusts the caller to implement NEP-11 correctly` is about.

> Explain about it a little more, let's say A particular token hash can be obtained from the transaction that is calling the contract.

Well, you get token hash via `System.Runtime.GetCallingScriptHash`, because NEP-11/NEP-17 contract will be doing this invocation. Transaction is a different matter.

> Transfer with multiple tokens?

One callback --- one token, it can't be any other way.

Reply:
"or in any other situation to
refer to contracts able to receive [[nep-11.mediawiki|NEP-11]] tokens (but not
necessarily being [[nep-11.mediawiki|NEP-11]] themselves)."

Reply:
Rephrased a bit, let's try it this way.